Mathlib/LinearAlgebra/Matrix/FixedDetMatrices.lean

```diff
@@ -52,4 +53,261 @@ lemma smul_coe (m : R) (g : SpecialLinearGroup n R) (A : FixedDetMatrix n R m) :
+  apply le_trans _ (Int.emod_lt (A.1 0 0) (abs_ne_zero.mp h))
+  simp only [Fin.isValue, Int.cast_id, add_le_add_iff_right]
+  rw [abs_eq_self.mpr (Int.emod_nonneg (A.1 0 0) (abs_ne_zero.mp h))]
+
+/--Reduction lemma for integral FixedDetMatrices. -/
+@[elab_as_elim]
+def reduce_rec {C : Δ m → Sort*} (h0 : ∀ A : Δ m, |(A.1 1 0)| = 0 → C A)
+    (h1 : ∀ A : Δ m, |(A.1 1 0)| ≠ 0 → C (reduce_step m A) → C A) :
+    ∀ A, C A := fun A => by
+  by_cases h : |(A.1 1 0)| = 0
+  · apply h0 _ h
+  · exact h1 A h (reduce_rec h0 h1 (reduce_step m A))
+  termination_by A => Int.natAbs (A.1 1 0)
+  decreasing_by
+    zify
+    apply reduce_aux m A
+    simpa only [Fin.isValue, ne_eq, abs_eq_zero] using h
+
+/--Map from `Δ m → Δ m` which reduces a FixedDetMatrix towards a representative element in reps. -/
+def reduce : Δ m → Δ m := fun A => by
+  if |(A.1 1 0)| = 0 then
+    if  0 < A.1 0 0 then exact (T ^ (-(A.1 0 1/A.1 1 1))) • A else exact
+      (T ^ (-(-A.1 0 1/ -A.1 1 1))) • ( S • ( S • A)) --the -/- don't cancel with ℤ divs.
+  else
+    exact reduce (reduce_step m A)
+  termination_by b => Int.natAbs (b.1 1 0)
+  decreasing_by
+      next a h =>
+      zify
+      apply reduce_aux m
+      simpa only [Fin.isValue, ne_eq, abs_eq_zero] using h
+
+lemma reduce_eqn1 (A : Δ m) (hc : |(A.1 1 0)| = 0) (ha : 0 < A.1 0 0) :
+    reduce m A = (T ^ (-(A.1 0 1/A.1 1 1))) • A := by
+  rw [reduce]
+  simp only [Fin.isValue, abs_eq_zero, zpow_neg, Int.ediv_neg, neg_neg, dite_eq_ite] at *
+  simp_rw [if_pos hc, if_pos ha]
+
+lemma reduce_eqn2 (A : Δ m) (hc : |(A.1 1 0)| = 0) (ha : ¬ 0 < A.1 0 0) :
+    reduce m A = (T ^ (-(-A.1 0 1/ -A.1 1 1))) • ( S • ( S • A)) := by
+  rw [reduce]
+  simp only [Int.cast_id, Fin.isValue, zpow_neg, Int.ediv_neg, neg_neg,
+    dite_eq_ite] at *
+  simp_rw [if_pos hc, if_neg ha]
+
+lemma reduce_eqn3 (A : Δ m) (hc : ¬ |(A.1 1 0)| = 0) :
+    reduce m A = reduce m (reduce_step m A) := by
+  rw [reduce]
+  simp only [Int.cast_id, Fin.isValue, zpow_neg, Int.ediv_neg, neg_neg,
+    dite_eq_ite] at *
+  simp_rw [if_neg hc]
+
+private lemma A_d_ne_zero (A : Δ m) (ha : A.1 1 0 = 0) (hm : m ≠ 0) : A.1 1 1 ≠ 0 := by
+  have := A.2
+  rw [det_fin_two, ha] at this
+  aesop
+
+private lemma A_a_ne_zero (A : Δ m) (ha : A.1 1 0 = 0) (hm : m ≠ 0) : A.1 0 0 ≠ 0 := by
+  have := A.2
+  rw [det_fin_two, ha] at this
+  aesop
+
+private lemma A_c_eq_zero (A : Δ m) (ha : A.1 1 0 = 0) : A.1 0 0 * A.1 1 1 = m := by
+  have := A.2
+  rw [det_fin_two, ha] at this
+  aesop
+
+/--An auxiliary result bounding the size of the entries of the representatives in `reps`. -/
+lemma reps_entries_le_m' (hm : m ≠ 0) (A : Δ m) (h : A ∈ reps m) (i j : Fin 2) :
+    A.1 i j ∈ Finset.Icc (-|m|) |m|:= by
+  have h1 : 0 < |A.1 1 1| := Eq.mpr (id (congrArg (fun _a ↦ _a) (propext abs_pos)))
+    (A_d_ne_zero m A h.left hm)
+  have h2 : 0 < |A.1 0 0| := Eq.mpr (id (congrArg (fun _a ↦ _a) (propext abs_pos)))
+    (A_a_ne_zero m A h.left hm)
```

Comment:
```suggestion
  have h1 : 0 < |A.1 1 1| := abs_pos.mpr (A_d_ne_zero m A h.left hm)
  have h2 : 0 < |A.1 0 0| := abs_pos.mpr (A_a_ne_zero m A h.left hm)
```
Where did these convoluted expressions come from?

Reply:
That was me using `show_term`